Deprecate defaulting to scalar in broadcast (take 2)

mbauman opened the pull request:
This PR replaces #26212.  The end result is the same, but I like this implementation much better.

This PR consists of two commits — both pass tests locally and I'd like to keep them separate. The first commit changes the default behavior of broadcast to collect its arguments instead of treating them as scalars. Broadcast now calls a special helper function, `broadcastable`, on each argument. It ensures that the arguments support indexing and have a shape, or otherwise have defined a custom `BroadcastStyle`.  The second commit layers on top a deprecation mechanism that makes this — almost entirely — non-breaking.

I believe this to be a compromise that will fix #18618 once deprecations are removed. Note that I kept all the types that are currently tested to behave as scalars as specializing `broadcastable` to return something that supports both indexing and `axes`, or otherwise has its own `BroadcastStyle` and internal implementation.

This also changes the behavior of broadcast such that if it ever wants to return a 0-dimensional array, it'll just "unwrap" that 0-dimensional array and return the contents instead.

Reasons I like this approach better than #26212:

* We're going to be requiring folks to explicitly wrap things in a `Ref` to make them broadcast as scalars.  In #26212, adding a `Ref` to an otherwise scalar expression would result in a 0-dimensional array as a result.  Yes, I generally have a very strong aversion to these kinds of auto-wrapping/unwrapping, but in this case I believe it is warranted and ends up with fewer special cases.  0-dimensional arrays and "scalar" types are treated identically as far as broadcast is concerned.
* Having an explicit `broadcastable` psuedo-conversion method makes it easier for other code to "work-around" broadcast and know how it'll behave.  Call `broadcastable` on a thing and then you'll be able to know what `broadcast` is going to do on the basis of its `axes`.  No digging around in `BroadcastStyle`s or other internals — this one function makes sure that iterables and everything else can just work.  (This isn't quite true, yet, as `Ref` isn't fully conforming with the iterable and `axes` spec, but I'd like to implement that separately.)
* It's so much simpler to not need to worry about `Broadcast.Scalar()` styles.

stevengj commented:
Does wrapping all scalars in a `Ref` affect performance?

mbauman commented:
Numbers have both `axes` and indexing, so they don't need to be wrapped in a `Ref`.  But even if you do wrap them in a ref, they're immediately unwrapped again (if all other arguments are 0-dim or similarly "scalar").

```
julia> f(a, b) = Ref(a) .+ Ref(b)
       g(a, b) = a .+ b
g (generic function with 1 method)

julia> @code_llvm f(1,2)

; Function f
; Location: REPL[7]:1
define i64 @julia_f_34293(i64, i64) {
top:
; Function broadcast; {
; Location: broadcast.jl:636
; Function broadcast; {
; Location: broadcast.jl:640
; Function +; {
; Location: int.jl:53
  %2 = add i64 %1, %0
;}}}
  ret i64 %2
}

julia> @code_llvm g(1,2)

; Function g
; Location: REPL[7]:2
define i64 @julia_g_34294(i64, i64) {
top:
; Function broadcast; {
; Location: broadcast.jl:228
; Function +; {
; Location: int.jl:53
  %2 = add i64 %1, %0
;}}
  ret i64 %2
}
```

vchuravy commented on base/deprecated.jl:
```diff
@@ -687,6 +687,20 @@ end
+    if Base.Broadcast.BroadcastStyle(typeof(x)) isa Broadcast.Unknown
+        depwarn("""
+            broadcast will default to iterating over its arguments in the future. Wrap arguments of
+            type `x::$(typeof(x))` with `Ref(x)` to ensure they broadcast as "scalar" elements.
```
Maybe add:

> or define `broadcastable(x::MyType) = Ref(x)` for the conversion to be automatic.

mbauman replied:
I'm guessing that the most common reader of this deprecation will be an end-user.

vchuravy commented on base/broadcast.jl:
```diff
@@ -432,6 +419,19 @@ end
+broadcastable(x::AbstractArray) = x
+# In the future, default to collecting arguments. TODO: uncomment once deprecations are removed
+# broadcastable(x) = BroadcastStyle(typeof(x)) isa Unknown ? collect(x) : x
+# broadcastable(::Dict) = error("intentionally unimplemented to allow development in 1.x")
```
Will this be a different PR? We should deprecate this behaviour in 0.7

mbauman replied:
`AbstractDict` hits the fallback `::Any` deprecation.  Next to that deprecation is a task-item to un-comment these definitions, so when that happens it will go from a warning to an error.

andyferris commented on base/broadcast.jl:
```diff
@@ -621,12 +630,17 @@ julia> string.(("one","two","three","four"), ": ", 1:4)
-@inline broadcast(f, A, Bs...) =
-    broadcast(f, combine_styles(A, Bs...), nothing, nothing, A, Bs...)
+@inline function broadcast(f, A, Bs...)
+    A′ = broadcastable(A)
```
It's unfortunate that the prime looks so similar to the adjoint...

mbauman replied:
¯\\_(ツ)_/¯

There's more than one of us using this style in base to represent roughly-the-same-value-but-maybe-slightly-different.

stevengj replied:
You could also use `Ã`.

andyferris commented on base/broadcast.jl:
```diff
@@ -432,6 +419,19 @@ end
+broadcastable(x::AbstractArray) = x
+# In the future, default to collecting arguments. TODO: uncomment once deprecations are removed
+# broadcastable(x) = BroadcastStyle(typeof(x)) isa Unknown ? collect(x) : x
+# broadcastable(::Dict) = error("intentionally unimplemented to allow development in 1.x")
```
`::AbstractDict`

andyferris replied:
Also we might consider if `NamedTuple` should maybe broadcast together with a `AbstractDict{Symbol}` rather than by iteration index.

vchuravy reviewed (commented):
I like this direction. Automatically unwrapping `Ref` is technically a breaking change, but I don't see a clean way of deprecating this. Maybe add a deprecation warning to `broadcastable(::Ref)`?

vchuravy commented:
@nanosoldier  `runbenchmarks(ALL, vs = ":master")`

mbauman commented:
Nanosoldier is going to fail without that commit I just pushed.

vtjnash commented on base/broadcast.jl:
```diff
@@ -102,15 +94,14 @@ behaves as an `N`-dimensional array for broadcasting. Specifically, `DefaultArra
 const DefaultMatrixStyle = DefaultArrayStyle{2}
 BroadcastStyle(::Type{<:AbstractArray{T,N}}) where {T,N} = DefaultArrayStyle{N}()
-BroadcastStyle(::Type{<:Ref}) = DefaultArrayStyle{0}()
+BroadcastStyle(::Type{<:Union{Base.RefValue,Number}}) = DefaultArrayStyle{0}()
```
Why is that a good thing (I don't actually know what this return value means)? For context, `RefArray` is a pointer to precisely one element of an array. It also changes `Ptr`, if that matters?

mbauman replied:
All arguments' `BroadcastStyle`s are combined with a promote-like mechanism in order to select a broadcasting implementation.  For the most part, it's largely used as a kludge around the lack of a "at least one vararg is of type `T`" dispatch, but it's more complicated to deal with cases that would otherwise be ambiguities.

Yes, `Ptr` has neither axes or indexing defined, so it cannot be — by itself — considered like a 0-dimensional array.  I don't think `RefArray` is common enough to add special broadcasting support for it.

Before:

```
julia> identity.(Ref([1,2,3],1))
0-dimensional Array{Int64,0}:
1
```

After:

```
julia> identity.(Ref([1,2,3],1))
┌ Warning: broadcast will default to iterating over its arguments in the future. Wrap arguments of
│ type `x::Base.RefArray{Int64,Array{Int64,1},Nothing}` with `Ref(x)` to ensure they broadcast as "scalar" elements.
│   caller = top-level scope
└ @ Core :0
Base.RefArray{Int64,Array{Int64,1},Nothing}([1, 2, 3], 1, nothing)
```

In the future, it will be an error because:

```
julia> collect(Ref([1,2,3],1))
ERROR: MethodError: no method matching length(::Base.RefArray{Int64,Array{Int64,1},Nothing})
```

vtjnash replied:
> I don't think RefArray is common enough to add special broadcasting support for it.

However, `RefValue` is an implementation detail, and should never be mentioned explicitly either: `RefArray` is simply an alternative implementation of it. We know that `length(::Ref) == 1` forall Ref subtypes, we've just hesitated to add that definition to the code until needed to keep the API small.

mbauman replied:
This is largely a historical accident due to how I ended up here… and trying to work around `Ptr`.  As it stands, I can easily remove all uses of `RefValue`.  I just pushed a change.

New after:

```
julia> identity.(Ref([1,2,3],1))
1
```

vchuravy commented:
Well it didn't listen to me anyway...

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

(Put I am just being overly cautious)

vtjnash commented on base/deprecated.jl:
```diff
@@ -687,6 +687,21 @@ end
 # Also un-comment the new definition in base/indices.jl
 
+# Broadcast no longer defaults to treating its arguments as scalar (#)
+function Broadcast.broadcastable(x)
```
should have `@noinline` to ensure depwarn tracking works

nalimilan commented on base/broadcast.jl:
```diff
@@ -432,6 +419,19 @@ end
+
+Return either `x` or an object like `x` such that it supports `axes` and indexing.
+"""
+broadcastable(x::Union{Symbol,String,Type,Function,UndefInitializer,Nothing,RoundingMode}) = Ref(x)
```
I guess you can add `Missing` to this list.

nalimilan commented on base/broadcast.jl:
```diff
@@ -457,13 +460,13 @@ as in `broadcast!(f, A, A, B)` to perform `A[:] = broadcast(f, A, B)`.
 end
 
 # Optimization for the all-Scalar case.
```
Is this still the "all scalar case"?

nalimilan commented on base/broadcast.jl:
```diff
@@ -549,17 +551,24 @@ combine_eltypes(f, A, As...) =
+arguments and the scalars themselves in `As`. Singleton and missing dimensions
+are expanded to match the extents of the other arguments by virtually repeating
+the value. By default, only a limited number of types are considered scalars,
+including `Number`s, `Type`s and `Function`s; all other arguments are iterated
```
Could give the full list (in particular strings).

nalimilan commented on base/broadcast.jl:
```diff
@@ -549,17 +551,24 @@ combine_eltypes(f, A, As...) =
+
+The resulting container type is established by the following rules:
 
- - If all the arguments are scalars, it returns a scalar.
```
I imagine this rule still applies? Maybe worth keeping for clarity even if it's covered by more general rules below.

nalimilan commented on NEWS.md:
```diff
@@ -905,6 +905,11 @@ Deprecated or removed
   * `map` on dictionaries previously operated on `key=>value` pairs. This behavior is deprecated,
     and in the future `map` will operate only on values ([#5794]).
 
+  * Previously, broadcast defaulted to treating its arguments as scalars if they were not
```
It would be useful to add a sentence for package developers mentioning `broadcastable`, just like the deprecation does.

nalimilan reviewed (commented):
Thanks for your efforts to improve `broadcast`!

Cc: @timholy

JeffBezanson commented on base/broadcast.jl:
```diff
@@ -432,6 +419,19 @@ end
+
+Return either `x` or an object like `x` such that it supports `axes` and indexing.
+"""
+broadcastable(x::Union{Symbol,AbstractString,Type,Function,UndefInitializer,Nothing,RoundingMode,Missing}) = Ref(x)
```
Can remove `Type` here since it has a method below.

mbauman committed:
Default to iterating over broadcast arguments

Broadcast now calls a special helper function, `broadcastable`, on each argument. It ensures that the arguments support indexing and have a shape, or otherwise have defined a custom `BroadcastStyle`.

mbauman committed:
Add a deprecation and comment out future behavior. Add NEWS and docs.

KristofferC commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/26435#issuecomment-372967193):
```
IOError(EOFError() during request(https://api.github.com/repos/JuliaLang/julia/statuses/fb59643d8cfa5011d66a4d09d8e38227b977a3e5))

```
cc @ararslan

fredrikekre commented on base/broadcast.jl:
```diff
@@ -432,6 +419,19 @@ end
+    broadcastable(x)
+
+Return either `x` or an object like `x` such that it supports `axes` and indexing.
+"""
```
Perhaps add an example?
````
```jldoctest
julia> Broadcast.broadcastable([1, 2, 3])
3-element Array{Int64,1}:
 1
 2
 3

julia> Broadcast.broadcastable("Hello, world!")
Base.RefValue{String}("Hello, world!")
```
````

stevengj commented on base/broadcast.jl:
```diff
@@ -423,6 +423,23 @@ end
 
 Return either `x` or an object like `x` such that it supports `axes` and indexing.
+
+If `x` supports iteration, the returned value should match [`collect(x)`](@ref).
```
This isn't true.  e.g. `collect(3)` returns an `Array{Int,0}`, but `broadcastable` just returns `3`.

mbauman replied:
Fair.  How about I add `with respect to axes and indexing`?

mbauman commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/26435#issuecomment-374321319) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/e16e9dea4843b98b4377bef760452283d49a4f2c/1263492_vs_b4af81d/report.md). cc @ararslan

StefanKarpinski commented:
That does not look too bad – all these regressions seem like they could be in the noise.

stevengj commented:
The slowdowns for tuple broadcasting seem like they might be real.  They are testing:
```
round.(Int, (rand(n)...,))
```
for various `n ≤ 10`.   Would be worth trying that manually to verify.

mbauman commented:
Well that was a remarkably simple fix.  Once CI passes again I'd like to merge.

vchuravy commented:
While we wait on CI.
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

ararslan commented:
> While we wait on CI.

Except that Nanosoldier takes about 3 times longer than our longest running CI build. :wink:

ararslan commented:
I had to restart the server. @nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/26435#issuecomment-374714684) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6635b562456fbadacba34590fa0c4768341ddf23/eca9dfb_vs_2723d74/report.md). cc @ararslan

vchuravy commented:
There is still a regression for `["broadcast", "typeargs", "(\"tuple\", 10)"]` and a deprecation warning in the Pkg3 tests

mbauman commented:
I fought with that remaining allocation for quite some time with no success.  It's the `Ref{Type{Int}}(Int)` that is getting heap-allocated in some situations, but it's not clear to me why.  Everything downstream of the `Ref` is inlined and it doesn't escape, but I think things are just on the edge of what the optimizer can see.  Nothing is obviously problematic to my eyes.  My attempts at reducing this down to a MWE have failed — it works as I'd expect once I pull the relevant methods out of Base and into my own module.  I'm guessing that it's because I've not copied the entire method tables.

Given that this regression is small and the optimizer is getting revamped, I'm inclined to merge.

mbauman committed:
fixup Pkg3 broadcasting over iterables

StefanKarpinski commented:
Would be interesting to see if the new optimizer can eliminate that allocation.

fredrikekre commented:
The manual section on the broadcasting interface still uses `Broadcast.Scalar` in a couple of places resulting in 
```
WARNING: Broadcast.Scalar is deprecated, use DefaultArrayStyle{0} instead.
  likely near /home/fredrik/julia-master/doc/make.jl:149
```
when building the docs.

fredrikekre commented:
And https://github.com/JuliaLang/julia/blob/7eb5d44a0caee9e7e13a956312c7ab10fbf09061/base/broadcast.jl#L593-L596 returns a tuple, but that seems to be intended.
```julia
julia> broadcast(+, 1.0, (0, -2.0), Ref(1))
(2.0, 0.0)
```
That example seem rather pointless now that `Ref(1)` and `1` behaves the same.

mbauman commented:
Thank you! See #26601.

fredrikekre commented:
There is a default, which is `broadcastable(x) = x`.

GiggleLiu commented:
@fredrikekre 
see the discuss here,

https://discourse.julialang.org/t/broadcasting-over-structs-in-1-0-what-am-i-supposed-to-do/13408

We have to implement the interface like
```julia
julia> Broadcast.broadcastable(m::M) = Ref(m)
```
to make a type work as a scalar.
